### ISSUE TITLE
Redeploy fails when container is in 'restarting' state

### ISSUE THREAD

**Issue by CVTJNII** [chef-cookbooks/docker#679]
During convergence there appears to be a race caused by containers in a restarting state.  v2.5 of the cookbook has been intermittently throwing the following error when attempting to redeploy a container:

```
             ================================================================================
             Error executing action `redeploy` on resource 'docker_container[InfluxDB]'
             ================================================================================
             
             Docker::Error::ConflictError
             ----------------------------
             Conflict, You cannot remove a running container. Stop the container before attempting removal or use -f
             
             Cookbook Trace:
             ---------------
             /tmp/kitchen/cache/cookbooks/docker/libraries/docker_container.rb:366:in `block (3 levels) in <class:DockerContainer>'
             /tmp/kitchen/cache/cookbooks/docker/libraries/helpers_base.rb:66:in `with_retries'
             /tmp/kitchen/cache/cookbooks/docker/libraries/docker_container.rb:366:in `block (2 levels) in <class:DockerContainer>'
             /tmp/kitchen/cache/cookbooks/docker/libraries/docker_container.rb:365:in `block in <class:DockerContainer>'
             /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/provider.rb:122:in `instance_eval'
             /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/provider.rb:122:in `compile_and_converge_action'
             /tmp/kitchen/cache/cookbooks/docker/libraries/docker_container.rb:159:in `call_action'
             /tmp/kitchen/cache/cookbooks/docker/libraries/docker_container.rb:356:in `block in <class:DockerContainer>'
             /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/provider.rb:122:in `instance_eval'
             /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/provider.rb:122:in `compile_and_converge_action'
             /tmp/kitchen/cache/cookbooks/monitoring_infra/recipes/influxdb.rb:91:in `block (2 levels) in from_file'
```

Dropping in some debug prints it looks like the container fails to stop as it is restarting:

```
           * docker_container[InfluxDB] action redeploy[2016-03-09T19:46:56+00:00] WARN: DEBUG: Stopping container with state {"Status"=>"restarting", "Running"=>true, "Paused"=>false, "Restarting"=>true, "OOMKilled"=>false, "Dead"=>false, "Pid"=>0, "ExitCode"=>1, "Error"=>"", "StartedAt"=>"2016-03-09T19:46:56.078352629Z", "FinishedAt"=>"2016-03-09T19:46:56.217049712Z"}
       
             - stopping InfluxDB  (will kill after 10s)[2016-03-09T19:46:56+00:00] WARN: DEBUG: Deleting container with state {"Status"=>"running", "Running"=>true, "Paused"=>false, "Restarting"=>false, "OOMKilled"=>false, "Dead"=>false, "Pid"=>6773, "ExitCode"=>0, "Error"=>"", "StartedAt"=>"2016-03-09T19:46:56.693474216Z", "FinishedAt"=>"2016-03-09T19:46:56.217049712Z"}
```

Print locations:

```
diff --git a/libraries/docker_container.rb b/libraries/docker_container.rb
index 437324a..43750d9 100644
--- a/libraries/docker_container.rb
+++ b/libraries/docker_container.rb
@@ -295,6 +295,7 @@ module DockerCookbook
     end
 
     action :stop do
+      Chef::Log.warn("DEBUG: Stopping container with state #{state}") 
       return unless state['Running']
       kill_after_str = " (will kill after #{kill_after}s)" if kill_after != -1
       converge_by "stopping #{container_name} #{kill_after_str}" do
@@ -360,6 +361,7 @@ module DockerCookbook
       return unless current_resource
       call_action(:unpause)
       call_action(:stop)
+      Chef::Log.warn("DEBUG: Deleting container with state #{state}") 
       converge_by "deleting #{container_name}" do
         with_retries { container.delete(force: force, v: remove_volumes) }
       end
```

I'm seeing this with Docker 1.10.0 and 1.10.2.

**Comment by CVTJNII** [comment]
This appears to be a bug in [helpers_container on line 78](https://github.com/chef-cookbooks/docker/blob/master/libraries/helpers_container.rb#L78).  The logic in wait_running_state is currently to block until state['Running'] matches the argument or until state['FinishedAt'] is not equal to '0001-01-01T00:00:00Z'.  However, in a restarting state Running is true, Restarting (not checked) is true, and FinshedAt is set.  This causes wait_running_state to return while the container is still stopping, leading to a race.  On redeploy, if the container stops before delete is called then things work, if it hasn't stopped yet then the cookbook throws an exception.

**Comment by chasebolt** [comment]
Good find! Can you submit a PR for this fix and a test (if possible)?

**Comment by CVTJNII** [comment]
Working on it.  The FinshedAt logic appears to be added in e079d777, currently trying to determine the spirit of this change so I don't cause a regression on another front.

**Comment by chasebolt** [comment]
That change was added in case a container was in a stopped state and will never be considered Running  again.

Here is more background info on it. https://github.com/chef-cookbooks/docker/issues/510

**Comment by CVTJNII** [comment]
So I actually uncovered another, much harder to reproduce race.  Occasionally, even with the correct logic in place, the container will fail to stop when it is restarting.  This appears to be something in Docker as the stop method completes successfully but the restarting container does not stop.  I believe I've worked around this with the following patches:

https://github.com/CVTJNII/docker/commit/ededfcfbd811860bb5b80ba62db6884004939038
https://github.com/CVTJNII/docker/commit/7df9203cfda1a374e63b37819eba7fad34caaf22

I don't have a test case for this unfortunately as it appears to be a container state race inside Docker, my testing was a painful and slow process of wrapping kitchen in a while loop and waiting for it to bomb.

I'm waiting to open PRs for these until we have a CCLA in place.
